packages/exhibitor/extra/dcos_zk_backup.py

```diff
@@ -0,0 +1,322 @@
+#!/usr/bin/env python3
+"""
+Command line tool for automated DC/OS ZooKeeper instance backup and restore.
+
+This script is made available as `dcos-zk` executable in DC/OS.
+
+"""
+import shlex
+import shutil
+import subprocess
+import sys
+import tarfile
+from argparse import ArgumentParser, ArgumentTypeError
+from pathlib import Path
+from tempfile import TemporaryDirectory
+
+
+EXHIBITOR_DIR = Path('/var/lib/dcos/exhibitor')
+
+
+def run_command(cmd: str, verbose: bool) -> None:
+    """
+    Run a command in a subprocess.
+
+    Args:
+        verbose: Show the output.
+
+    Raises:
+        subprocess.CalledProcessError: The given cmd exits with a non-0 exit code.
+    """
+    stdout = None if verbose else subprocess.PIPE
+    stderr = None if verbose else subprocess.STDOUT
+    subprocess.run(
+        args=shlex.split(cmd),
+        stdout=stdout,
+        stderr=stderr,
+        check=True,
+    )
+
+
+def _copy_dir_and_preserve_ownership(src: Path, dst: Path, verbose: bool) -> None:
+    """
+    Copy a directory from ``src`` to ``dst`` and preserve ownership.
+
+    We need to preserve ownership for a successful restore.
+    We use a subprocess (with `cp -p`) for this rather than Python's
+    `shutil.copytree` as `copytree` does not preserve ownership.
+    """
+    run_command('cp -prv {src} {dst}'.format(src=src, dst=dst), verbose)
+
+
+def backup_zookeeper(
+    backup: Path,
+    tmp_dir: Path,
+    verbose: bool,
+) -> None:
+    """
+    DC/OS ZooKeeper instance backup procedure.
+
+    ZooKeeper changes files while it is running.
+    In order to have a consistent backup stop ZooKeeper, run the backup procedure, then start ZooKeeper.
+
+    Running this procedure will temporarily bring down the ZooKeeper instance on this master node.
+    This means if the DC/OS cluster currently operates on a quorum of int(N / 2) + 1, where N is the
+    expected number of DC/OS master nodes, the quorum will be lost while the backup is taking place
+    and the cluster will experience down time.
+
+    See https://jira.mesosphere.com/browse/DCOS_OSS-5185
+    for changing this procedure to allow a backup without downtime.
+
+    If backing up ZooKeeper via a custom temporary directory, a failing backup procedure does not
+    clean up files created in the temporary directory in the process.
+    """
+    zookeeper_dir = EXHIBITOR_DIR / 'zookeeper'
+    tmp_zookeeper_dir = tmp_dir / 'zookeeper'
+
+    print('Backing up ZooKeeper into {backup} via {tmp_zookeeper_dir}'.format(
```

Comment:
I would replace `print` statements with `logging` module use.

Reply:
Good point. I consulted the Python3 documentation and then decided to use `print` instead of logging: https://docs.python.org/3/howto/logging.html#when-to-use-logging
It seems reasonable to me to distinguish the two, `print` as a CLI user interface, logging as a debugging help for developers.